src/pyk/kcfg/store.py

```diff
@@ -0,0 +1,133 @@
+from __future__ import annotations
+
+import threading
+from abc import abstractmethod
+from collections.abc import MutableMapping
+from dataclasses import dataclass, field
+from typing import TYPE_CHECKING, Generic, TypeVar, final
+
+from ..cterm import CTerm
+from ..kast.inner import KApply, KSequence, KToken, KVariable, bottom_up_with_summary
+from .kcfg import KCFG
+
+if TYPE_CHECKING:
+    from collections.abc import Iterator
+
+    from ..kast.inner import KInner, KLabel
+
+
+A = TypeVar('A')
+
+
+@dataclass
+class CachedValues(Generic[A]):
+    value_to_id: dict[A, int] = field(default_factory=dict)
+    values: list[A] = field(default_factory=list)
+
+    def cache(self, value: A) -> int:
+        id = self.value_to_id.get(value)
+        if id is not None:
+            return id
+        id = len(self.values)
+        self.value_to_id[value] = id
+        self.values.append(value)
+        return id
+
+
+@dataclass(eq=True, frozen=True)
```

Comment:
`eq` is `True` by default:

```suggestion
@dataclass(frozen=True)
```